### PR TITLE
HIP-17: Update HTS with NFT suppport

### PR DESCRIPTION
**Pull request by Daniel-K-Ivanov**
**Related issue(s)**:
Rlated to: #1605 

**Summary of the change**:
More notable changes:
- Adds `SubType` for each `HederaFunctionality`, allowing for 2 `FeeData` structures per Hedera Operation
- `MerkleUniqueTokenId` and `MerkleUniqueToken` added representing the state data for Unique Tokens (NFTS)
- Adds new properties to `MerkleToken`; Added `TokenType` and `TokenSupplyType`
- Updates `TokenMint` to support `NFT` minting 
- Adds`TokenGetNftInfo` query
- Adds multiple `GlobalDynamicProperties`
- Adds temporary `FCInvertibleHashMap` used for Unique Token datastructures
- Updates `MerkleAccount` adding `ownedNfts`
- Adds new `nftTransfers` in `ExpirableTransactionRecord`

### CODE REVIEW

File: hapi-fees/src/main/java/com/hedera/services/usage/crypto/CryptoTransferUsage.java [hunk 1]
```diff
@@ -0,0 +1,77 @@
...
+import com.hedera.services.usage.SigUsage;
+import com.hedera.services.usage.TxnUsageEstimator;
+import com.hederahashgraph.api.proto.java.FeeData;
+import com.hederahashgraph.api.proto.java.TokenTransferList;
+import com.hederahashgraph.api.proto.java.TransactionBody;
+
+import static com.hedera.services.usage.SingletonEstimatorUtils.ESTIMATOR_UTILS;
+import static com.hederahashgraph.fee.FeeBuilder.BASIC_ENTITY_ID_SIZE;
+
+public class CryptoTransferUsage extends CryptoTxnUsage<CryptoTransferUsage> {
```
Comment (tinker-michaelj) [inline hapi-fees/src/main/java/com/hedera/services/usage/crypto/CryptoTransferUsage.java:32]: We should be able to delete this class; it is superseded by `CryptoOpsUsage.cryptoTransferUsage()` above.

File: hapi-fees/src/test/java/com/hedera/services/usage/crypto/CryptoTransferUsageTest.java [hunk 2]
```diff
@@ -0,0 +1,157 @@
...
+import static com.hedera.services.test.UsageUtils.A_USAGES_MATRIX;
+import static com.hedera.services.usage.SingletonUsageProperties.USAGE_PROPERTIES;
+import static com.hedera.services.usage.token.entities.TokenEntitySizes.TOKEN_ENTITY_SIZES;
+import static org.junit.Assert.assertEquals;
+import static org.mockito.BDDMockito.any;
+import static org.mockito.BDDMockito.given;
+import static org.mockito.BDDMockito.mock;
+import static org.mockito.BDDMockito.verify;
+
+public class CryptoTransferUsageTest {
```
Comment (tinker-michaelj) [inline hapi-fees/src/test/java/com/hedera/services/usage/crypto/CryptoTransferUsageTest.java:53]: TBDeleted

File: hedera-node/src/main/java/com/hedera/services/fees/calculation/UsagePricesProvider.java [hunk 3]
```diff
@@ -43,13 +45,34 @@
 	void loadPriceSchedules();
 
 	/**
-	 * Returns the prices in tinyCents that must be paid to
+	 * Returns the prices in a map SubType keys and FeeData values in tinyCents that must be paid to
```
Comment (tinker-michaelj) [inline hedera-node/src/main/java/com/hedera/services/fees/calculation/UsagePricesProvider.java:48]: I didn't notice for a while either, but the unit of `FeeData` prices is actually 1/1000th of a tinycent.

File: hedera-node/src/main/java/com/hedera/services/state/submerkle/NftAdjustments.java [hunk 4]
```diff
@@ -0,0 +1,132 @@
...
+			return false;
+		}
+		NftAdjustments that = (NftAdjustments) o;
+		return Arrays.equals(serialNums, that.serialNums)
+				&& senderAccIds.equals(that.senderAccIds) && receiverAccIds.equals(that.receiverAccIds);
+	}
+
+	@Override
+	public int hashCode() {
+		int result = Long.hashCode(RUNTIME_CONSTRUCTABLE_ID);
```
Comment (tinker-michaelj) [inline hedera-node/src/main/java/com/hedera/services/state/submerkle/NftAdjustments.java:85]: Not very important, but `hashCode()` is recommended to use only the fields that affect `equals()`.

File: hedera-node/src/main/java/com/hedera/services/store/TypedTokenStore.java [hunk 5]
```diff
@@ -209,6 +245,15 @@ public void persistToken(Token token) {
 		final var mutableToken = currentTokens.getForModify(key);
 		mapModelChangesToMutable(token, mutableToken);
 
+		if (token.hasMintedUniqueTokens()) {
+			for (UniqueToken uniqueToken : token.mintedUniqueTokens()) {
+				final var merkleUniqueId = new MerkleUniqueTokenId(new EntityId(uniqueToken.getTokenId()), uniqueToken.getSerialNumber());
+				final var merkleUniqueToken = new MerkleUniqueToken(new EntityId(uniqueToken.getOwner()), uniqueToken.getMetadata(), uniqueToken.getCreationTime());
+				uniqueTokens.get().put(merkleUniqueId, merkleUniqueToken);
+			}
+			accountStore.persistAccount(token.getTreasury());
```
Comment (tinker-michaelj) [inline hedera-node/src/main/java/com/hedera/services/store/TypedTokenStore.java:254]: I'm leaning toward a pattern in which a `TransitionLogic` must explicitly call a `*Store.persist*` method for **every** entity it expects to be changed in the transaction.

This does lead to some awkward idioms, but I like the single-method-visibility it gives to exactly what state changes will be caused by the transaction.

File: hedera-node/src/main/java/com/hedera/services/store/tokens/HederaTokenStore.java [hunk 6]
```diff
@@ -328,6 +341,43 @@ public ResponseCodeEnum adjustBalance(AccountID aId, TokenID tId, long adjustmen
...
 
+	@Override
+	public ResponseCodeEnum adjustBalance(AccountID senderAId, AccountID receiverAId, TokenID tId, long serialNumber) {
+		var validity = sanityChecked(senderAId, receiverAId, tId, token -> tryAdjustment(senderAId, receiverAId, tId, serialNumber));
+		if (validity == OK) {
+			var nftId = NftID.newBuilder().setTokenID(tId).setSerialNumber(serialNumber).build();
+			if (nftExists(nftId)) {
+				get(nftId).setOwner(EntityId.fromGrpcAccountId(receiverAId));
+
+				adjustOwnedNfts(senderAId, false);
```
Comment (tinker-michaelj) [inline hedera-node/src/main/java/com/hedera/services/store/tokens/HederaTokenStore.java:352]: This seems quite risky, as `adjustOwnedNfts()` is irreversibly changing the mutable leaf right here, while the surrounding code is just adding "provisional" changes in the `TransationalLedger` proxy---which in principle could be rolled back.

Probably a non-issue given the ongoing refactor, however.

File: hedera-node/src/main/java/com/hedera/services/txns/validation/TokenListChecks.java [hunk 7]
```diff
@@ -44,104 +46,145 @@
...
 import static com.hederahashgraph.api.proto.java.ResponseCodeEnum.INVALID_TOKEN_INITIAL_SUPPLY;
+import static com.hederahashgraph.api.proto.java.ResponseCodeEnum.INVALID_TOKEN_MAX_SUPPLY;
 import static com.hederahashgraph.api.proto.java.ResponseCodeEnum.INVALID_WIPE_KEY;
+import static com.hederahashgraph.api.proto.java.ResponseCodeEnum.NOT_SUPPORTED;
 import static com.hederahashgraph.api.proto.java.ResponseCodeEnum.OK;
 import static com.hederahashgraph.api.proto.java.ResponseCodeEnum.TOKEN_ID_REPEATED_IN_TOKEN_LIST;
 import static com.hederahashgraph.api.proto.java.ResponseCodeEnum.TRANSFERS_NOT_ZERO_SUM_FOR_TOKEN;
 
 public class TokenListChecks {
-    static Predicate<Key> ADMIN_KEY_REMOVAL = ImmutableKeyUtils::signalsKeyRemoval;
```
Comment (tinker-michaelj) [inline hedera-node/src/main/java/com/hedera/services/txns/validation/TokenListChecks.java:53]: The imminent merge of https://github.com/hashgraph/hedera-services/pull/1616 will cause serious conflicts here; I will work on resolving those in the `refactored-nfts-touchups` tomorrow.

**Review by tinker-michaelj — COMMENTED** [review]
Awesome work!! 

Tomorrow I will merge `01589-M-DoAtomicXfersWithModelIds` from https://github.com/hashgraph/hedera-services/pull/1616 into `refactored-nft-touchups`. (The `01589-M-DoAtomicXfersWithModelIds` branch has some critical refactoring needed for HIP-18.)

File: hedera-node/src/main/java/com/hedera/services/store/tokens/HederaTokenStore.java [hunk 8]
```diff
@@ -328,6 +341,43 @@ public ResponseCodeEnum adjustBalance(AccountID aId, TokenID tId, long adjustmen
 		return sanityChecked(aId, tId, token -> tryAdjustment(aId, tId, adjustment));
 	}
 
+	@Override
+	public ResponseCodeEnum adjustBalance(AccountID senderAId, AccountID receiverAId, TokenID tId, long serialNumber) {
```
Comment (tinker-michaelj) [inline hedera-node/src/main/java/com/hedera/services/store/tokens/HederaTokenStore.java:345]: Actually, where do we validate that the `senderAId` is the current owner of this `serialNumber`?

File: hapi-fees/src/main/java/com/hedera/services/usage/token/TokenBurnUsage.java [hunk 9]
```diff
@@ -36,15 +40,34 @@ public static TokenBurnUsage newEstimate(TransactionBody tokenBurnOp, SigUsage s
...
 		return this;
 	}
 
 	public FeeData get() {
+		var op = this.op.getTokenBurn();
+
+		if (currentSubType == SubType.TOKEN_NON_FUNGIBLE_UNIQUE) {
+			var bytesToAdd = 0;
+			for (Long o : op.getSerialNumbersList()) {
+				bytesToAdd += o.byteValue();
```
Comment (tinker-michaelj) [inline hapi-fees/src/main/java/com/hedera/services/usage/token/TokenBurnUsage.java:59]: ‼️ `o.byteValue()` could easily be negative---e.g. `Long.valueOf(37047).byteValue()`...should we just have
```
bytesToAdd = LONG_SIZE * op.getSerialNumbersList().size()
```
or something to that effect?

Comment (Daniel-K-Ivanov) [reply hapi-fees/src/main/java/com/hedera/services/usage/token/TokenBurnUsage.java:59]: The bug is resolved in an upcoming PR

File: test-clients/src/main/java/com/hedera/services/bdd/spec/queries/token/HapiGetAccountNftInfos.java [hunk 10]
```diff
@@ -0,0 +1,136 @@
...
+        return HederaFunctionality.TokenGetNftInfo;
+    }
+
+    @Override
+    protected HapiGetAccountNftInfos self() {
+        return this;
+    }
+
+    @Override
+    protected void assertExpectationsGiven(HapiApiSpec spec) throws Throwable {
```
Comment (Daniel-K-Ivanov) [inline test-clients/src/main/java/com/hedera/services/bdd/spec/queries/token/HapiGetAccountNftInfos.java:71]: There is a bug in the assertion. We have to make sure that the whole list is equal to the expected one. Atm we are searching for an Nft in the expected list and that has problems if we mint 2 different Nft (`1` and `2) and try to assert that the minted nfts are `1` and `1`

Comment (Daniel-K-Ivanov) [reply test-clients/src/main/java/com/hedera/services/bdd/spec/queries/token/HapiGetAccountNftInfos.java:72]: resolved with #1724

**Review by qnswirlds — APPROVED** [review]
Overall LGTM. Will do more refactoring and testing enhancements in another PR.

**Review by tinker-michaelj — APPROVED** [review]
🍾

**Pull request merged**